torch/distributed/algorithms/model_averaging/averagers.py

```diff
@@ -99,15 +100,31 @@ def __init__(
             )
         self.period = period
 
-    def average_parameters(self, params):
-        r"""
-        Averages parameters if ``step`` is no less than ``warmup_steps``
+    def average_parameters(self, params: Union[Iterable[torch.nn.Parameter], Iterable[Dict[str, torch.nn.Parameter]]]):
+        """
+        Args:
+            params: average model.parameters() or parameter groups of an optimizer
+        Averages parameters or parameter groups of an optimizer if ``step`` is no less than ``warmup_steps``
         and it can be divided by ``period``, where ``step`` is increased by 1
         at each iteration in the training loop.
         """
         if (
             self.step >= self.warmup_steps
             and (self.step - self.warmup_steps) % self.period == 0
         ):
-            utils.average_parameters(iter(params), self.process_group)
+            filter_params = []
```

Comment:
Refactoring suggestion:

1. Wrap L115-L128 into a util function `get_params_to_average` in `torch/distributed/algorithms/model_averaging/utils.py`.
2. Create a util function `average_parameters_or_parameter_groups` in `torch/distributed/algorithms/model_averaging/utils.py`, which calls `get_params_to_average` and `average_parameters`.
3. Use  `average_parameters_or_parameter_groups` here and `hierarchical_model_averager.py`.

Reply:
the new function is following:

def average_parameters_or_parameter_groups(params: Union[Iterable[torch.nn.Parameter], Iterable[Dict[str, torch.nn.Parameter]]], process_group: dist.ProcessGroup):
    """
    average parameters or parameter_groups
    """
    filter_params = get_params_to_average(params)
    average_parameters(filter_params, process_group)

since filter_params is a list, so no need to iter filter_params.
so i change params type from Iterator[torch.nn.Parameter] to List[torch.nn.Parameter].
Of course the params argument can accept Iterator[torch.nn.Parameter] input.
so what is your suggest? keep Iterator or List ?